corehq/apps/user_importer/importer.py

```diff
@@ -683,6 +686,16 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
             status_row['flag'] = e.message
         except (UserUploadError, CouchUser.Inconsistent) as e:
             status_row['flag'] = str(e)
+        except Exception as e:
+            # HACK: Catching all exception here is temporary. We believe that user critical sections
+            # are not behaving properly, and this catch-all is here to identify the problem
+            status_row['flag'] = str(e)
+            notify_error(f'Error while processing bulk import: {str(e)}')
+            soft_assert(to='{}@{}'.format('mriley', 'dimagi.com'), send_to_ops=False)(
+                False,
+                'Error while processing bulk import',
+                e
+            )
```

Comment:
This is pretty far out for such a generic exception, given how deep the new potential exception is. I get your caution—you're lifting a except Exception deep inside that utility and there's no telling what demons that releases.

On the other hand, if you think that adding this broad handling of unforeseen errors is actually just a better way of dealing with _any_ kind of unforeseen errors in this bulk upload code, then I might change the expectation there that this is temporary.

Not blocking, just food for thought. If you like it how it is, I'm okay with it.

Reply:
@dannyroberts Updated the PR -- is scoping the exception this way more what you had in mind? Unfortunately this introduces some duplication. Unfortunately, I think the code is in need of a larger refactor to make the change clean

Reply:
Does the `fail_hard` param passed to `CriticalSection` mean that there's a base redis exception that can be caught?  `user.save()` does quite a lot of things.

Also I'm wondering about the `continue` here - does that mean that imports will see partial successes?  I'm guessing right now there are categories of errors where we reject the whole import with no changes, and others where we do partial successes, is that right?  The failure mode before this change probably would've been to save all users up to the erroring user and none after, probably with some cryptic failure notification, and after this change the failure mode is to skip the erroring user(s) and add a warning.  That seems reasonable to me, just want to be sure that change is in line with expectations.

Reply:
`fail_hard` just propagates whatever `Exception` it ran into when trying to obtain the `CriticalSection`: https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/dimagi/utils/couch/__init__.py#L303-L305. While I had theorized (and was able to replicate) that Redis going down would lead to the observed behavior, we didn't see any confirmation that this was actually what happened. And, at least from trying to run HQ with Redis down, it seems a lot of things would have failed if that were the case. So -- I'm confident that we are obtaining the `CriticalSection` when we shouldn't be, but I'm not sure why that is happening. Because of that, we can't be any more specific about what kind of `Exception` we are catching -- and that is why this fix is intended to be temporary until we gather the specific root cause.

Handling this error doesn't change the behavior of the importer -- it did partial success before as well. The [code](https://github.com/dimagi/commcare-hq/blob/fa8be9daf7986d3e205ae229f8b5915b94de3e30/corehq/apps/user_importer/importer.py#L682C1-L687C39) was already catching `ValidationError`, `UserUploadError`, and `CouchUser.Inconsistent`. If the code caught any of those errors, an failure entry would just be generated and the upload would continue. I had initially just caught the generic `Exception` here, but Danny pointed out that doing so broadened the failures we were catching beyond just a `CriticalSection` failure.

Reply:
> it did partial success before as well

Ah I see - I had assumed that wrapped the whole set of users, but I see that it is indeed per-user.  I don't want to get too involved in figuring out the best course here, mostly just wanted to flag the unintuitively large set of things that happen inside `user.save()` in case you hadn't looked at those methods/signals before.  I'm still haunted by a bug I saw once where user.save() called something that called something that called user.save(), so we hit an infinite recursion error, but that was swallowed silently by `.send_robust`

Reply:
Thanks for bringing that up. Given that the importer already seems to want to handle errors per-user without disrupting the entire process, I feel comfortable with the current approach of catching all `Exceptions` that `user.save()` could throw here. Again, I'd like to restrict the scope of this once we can detect what actual exception is occurring.